engine/timer_task_queue.go

```diff
@@ -0,0 +1,295 @@
+			w.currMaxLoadedTaskSequence = resp.MaxSequenceInclusive
+
+			w.remainingToFireTimersHeap = NewTimerTaskPriorityQueue(resp.Tasks)
+
+			task0 := w.remainingToFireTimersHeap[0]
+			w.nextFiringTimer.Update(time.Unix(task0.FireTimestampSeconds, 0))
+		}
+	}
+}
+
+func (w *timerTaskQueueImpl) drainAllNotifyRequests(initReq *xdbapi.NotifyTimerTasksRequest) []xdbapi.NotifyTimerTasksRequest {
+	var reqs []xdbapi.NotifyTimerTasksRequest
+	if initReq == nil {
+		reqs = make([]xdbapi.NotifyTimerTasksRequest, 0, len(w.triggeredPollingChan))
+	} else {
+		reqs = make([]xdbapi.NotifyTimerTasksRequest, 0, len(w.triggeredPollingChan)+1)
+		filteredReq := w.filterNotifyRequest(*initReq)
+		if filteredReq != nil {
+			reqs = append(reqs, *filteredReq)
+		}
+	}
+
+	for len(w.triggeredPollingChan) > 0 {
+		req, ok := <-w.triggeredPollingChan
+		if ok {
+			filteredReq := w.filterNotifyRequest(req)
+			if filteredReq != nil {
+				reqs = append(reqs, *filteredReq)
+			}
+		}
+	}
+	return reqs
+}
+
+func (w *timerTaskQueueImpl) shouldLoadNextBatch() bool {
+	if (!w.nextPreloadTimer.IsActive() && // this means the preload timer has fired
+		len(w.remainingToFireTimersHeap) == 0) && // this means all the timer tasks in the heap have been fired and sent to processor
+		len(w.firedToCompleteTimerSequenceMap) == 0 { // this means all the fired timers have been completed
+		return true
+	}
+	return false
+}
+
+func (w *timerTaskQueueImpl) sendFiredTimerToProcessor() {
+	for {
+		if len(w.remainingToFireTimersHeap) == 0 {
+			break
+		}
+		minTask := w.remainingToFireTimersHeap[0]
+		if minTask.FireTimestampSeconds <= time.Now().Unix() {
+			heap.Pop(&w.remainingToFireTimersHeap)
+			w.processor.GetTasksToProcessChan() <- *minTask
+		} else {
+			w.nextFiringTimer.Update(time.Unix(minTask.FireTimestampSeconds, 0))
+			break
+		}
+	}
+}
+
+func (w *timerTaskQueueImpl) triggeredPolling(reqs []xdbapi.NotifyTimerTasksRequest) {
+	if len(reqs) == 0 {
+		return // nothing to poll
+	}
+
+	resp, err := w.store.GetTimerTasksForTimestamps(
+		w.rootCtx, persistence.GetTimerTasksForTimestampsRequest{
+			ShardId:              w.shardId,
+			MinSequenceInclusive: w.currMaxLoadedTaskSequence + 1,
+			DetailedRequests:     reqs,
+		})
+
+	if err != nil {
+		w.logger.Error("failed at triggered polling timer task, will not retry. "+
+			"The new timers will be waiting for next preload", tag.Error(err))
+		// Give up on error as this notification mechanism is an optimization.
+		// The next preload will poll them anyway
+	} else {
+		if len(resp.Tasks) > 0 {
+			// update the max loaded sequence so that next time it won't load the same tasks
+			// currMaxLoadedTaskTimestamp is not updated because the new tasks won't have a bigger timestamp
+			w.currMaxLoadedTaskSequence = resp.MaxSequenceInclusive
+
+			// add the new tasks into the heap
+			for _, task := range resp.Tasks {
+				heap.Push(&w.remainingToFireTimersHeap, &task)
```

Comment:
Will it push duplicated tasks into the heap?

Reply:
that's the idea of tracking `currMaxLoadedTaskSequence`, and use it for every `GetTimerTasksForTimestamps`. Because the taskSequence is always increasing meaning the new tasks will also have a higher sequence number. The dup tasks will be filtered out when selecting it.